### PR TITLE
KafkaTopicDataObject: fix passing through tombstone messages (#915)

### PR DESCRIPTION
**Pull request by zzeekk**
see #915

### CODE REVIEW

File: sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala [hunk 1]
```diff
@@ -349,11 +349,11 @@ case class KafkaTopicDataObject(override val id: DataObjectId,
 
   private def convertToWriteDataFrame(df: DataFrame): DataFrame = {
     require(df.columns.toSet == Set("key","value"), s"($id) Expects columns key, value in DataFrame for writing to Kafka. Given: ${df.columns.mkString(", ")}")
-    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "read (keySchema)"))
-    valueSchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("value").dataType.asInstanceOf[StructType]), "read (valueSchema)"))
+    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "write (keySchema)"))
```
Comment (zzeekk) [inline sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala:355]: note: this is just a cosmetic change for a log statement...

File: sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala [hunk 2]
```diff
@@ -347,21 +349,21 @@ case class KafkaTopicDataObject(override val id: DataObjectId,
...
+  private def encodeKeyValue(df: DataFrame): DataFrame = {
     require(df.columns.toSet == Set("key","value"), s"($id) Expects columns key, value in DataFrame for writing to Kafka. Given: ${df.columns.mkString(", ")}")
-    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "read (keySchema)"))
-    valueSchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("value").dataType.asInstanceOf[StructType]), "read (valueSchema)"))
+    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "write (keySchema)"))
+    valueSchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("value").dataType.asInstanceOf[StructType]), "write (valueSchema)"))
     df.select(
       convertToKafka(keyType, col("key"), SubjectType.key, keySchema).as("key"),
-      convertToKafka(valueType, col("value"), SubjectType.value, valueSchema).as("value")
+      when(col("value").isNotNull, convertToKafka(valueType, col("value"), SubjectType.value, valueSchema)).as("value")
```
Comment (iliassoto) [inline sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala]: When decoding the value in *decodeKeyValue*, we exclude values that are NULL **and** the ones that have length == 0. Here, we only convert nonNull values to Kafka. Why the difference?

**Review by iliassoto — COMMENTED** [review]
See one comment. LGTM otherwise

File: sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala]: Good question.
In decodeKeyValue a Binary column has to be decoded. If its value is an empty binary array, decoding probably fails.
In encodeKeyValue it is a column of arbitrary type. Checking for length is possible for Array, String and Binary, but is not defined for Struct or Numeric types.
To have the same definition we could check length=0 if it is a binary type. I have already adapted the code.

**Comment by zzeekk** [comment]
@iliassoto: comment resolved, please approve again.
Note: we have changed a setting recently so that PRs can not be merge anymore if there are open conversations. So next time you can make comments to resolve, and approve the PR already, if you agree that we merge after all comments have been resolved...

**Review by iliassoto — APPROVED** [review]

**Pull request merged**